lib/benchee/benchmark/measure/memory.ex

```diff
@@ -52,25 +57,26 @@ defmodule Benchee.Benchmark.Measure.Memory do
     end)
   end
 
-  defp return_memory({memory_usage, result}) when memory_usage < 0, do: {nil, result}
-  defp return_memory({memory_usage, result}), do: {memory_usage, result}
+  defp return_memory({memory_usage, return_value}) when memory_usage < 0, do: {nil, return_value}
+  defp return_memory(memory_usage_info), do: memory_usage_info
 
-  defp measure_memory(fun, tracer, parent) do
+  defp measure_memory(fun, tracer) do
     :erlang.garbage_collect()
     send(tracer, :begin_collection)
 
     receive do
       :ready_to_begin -> nil
     end
 
-    result = fun.()
-    send(parent, {:result, result})
+    fun.()
```

Comment:
is this worth it? I'm not sure but don't we sort of get rid off the memory of the return value? I think that's fair to include. I'm uncertain though.

Reply:
It makes a significant difference for small benchmarks. For example, the first test case we have right now on `master` reports to use 440 bytes, but with this patch reports 368 bytes. For bigger benchmarks it's not such a big deal, but that's a big variance for smaller cases.

Plus, I'm not so worried about having a smaller sample size for memory measurements since they're almost always identical. That's the real side effect of running the function twice.

Reply:
I'm not _super_ worried about the sample size but once we leave micro benchmarks it produces interesting problems. My function takes 3 seconds to execute. I'm sure it's static so I set a memory time of 3 seconds - but it takes 6 seconds. That's weird behaviour.

I'm also not sure - technically the result matters as for instance when I want to benchmark the memory usage of a map operation the produced list obviously should count. I'm not sure if that's what's happening or if having a variable by itself is already a memory overhead (shouldn't be) - hence I'm wondering if it's "right" to remove this as well additionally to my concerns.

Reply:
We're definitely still counting the final term returned from the function in our measurements. What this removes is the need to allocate an extra tuple for `{:result, result}` that wasn't part of the function we're measuring. We're currently counting that in our measurements, which is not accurate.

That's a good point, though, about bigger functions. I would initially think that we could the benchmark and the function to get the return value concurrently, but that might cause unexpected behavior for a good number of functions if the user isn't expecting the functions to be run concurrently, so I think that idea's out.

We definitely need the return value of the function for the `after` hooks, so we can't just return `nil` or something. Maybe we can keep what we have (sending the result to the parent process) and just adjust the final number by subtracting the memory used by that extra tuple? The only problem there is that isn't a static number - it varies from platform to platform.

What we could do is to `send(parent, result)`. We could change the `receive` block on the parent to be:

```elixir
    receive do
      {^ref, memory_usage} ->
        return_memory({memory_usage, result})

      {^ref, :shutdown} ->
        nil

      new_result ->
        await_results(new_result, ref)
    end
```

It's a little dangerous assuming any message sent to the parent that doesn't have the reference in there is our result, but based on what we have now, that should work, and it wouldn't allocate anything extra. I'll poke around and see how this works.

Reply:
Ok, I've pushed up an extra commit that makes it so we don't need to run the function twice, and we still get the correct measurement of 368 bytes for our most simple test case in our `memory_test.exs` test. Let me know what you think!